core-java-modules/core-java-concurrency-2/src/test/java/com/baeldung/concurrent/completablefuture/CombiningCompletableFuturesUnitTest.java

```diff
@@ -0,0 +1,105 @@
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.junit.Test;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.mockito.Mock;
+
+public class CombiningCompletableFuturesUnitTest {
+
+    @Mock Logger logger;
+
+    @BeforeEach
+    void setup() {
+        logger = mock(Logger.class);
+    }
+
+    private static Stream<Arguments> clientData() {
+        return Stream.of(
+            Arguments.of(List.of("Good Resource"), 1, 0),
+            Arguments.of(List.of("Bad Resource"), 0, 1),
+            Arguments.of(List.of("Good Resource", "Bad Resource"), 1, 1),
+            Arguments.of(List.of("Good Resource", "Bad Resource", "Good Resource", "Bad Resource", "Good Resource"), 3, 2)
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("clientData")
+    public void givenMicroserviceClient_whenMultipleCreateResource_thenCombineResults(List<String> inputs, int successCount, int errorCount) {
+        MicroserviceClient mockMicroservice = mock(MicroserviceClient.class);
+        // Return an identifier of 123 on "Good Resource"
+        when(mockMicroservice.createResource("Good Resource"))
+            .thenReturn(CompletableFuture.completedFuture(123L));
+        // Throw an exception on "Bad Resource"
+        when(mockMicroservice.createResource("Bad Resource"))
+            .thenReturn(CompletableFuture.failedFuture(new IllegalArgumentException("Bad Resource")));
+
+        // Given a list of CompletableFutures from our microservice calls...
+        List<CompletableFuture<Long>> clientCalls = new ArrayList<>();
+        for (String resource : inputs) {
+            clientCalls.add(mockMicroservice.createResource(resource));
+        }
+
+        // When all CompletableFutures are completed (exceptionally or otherwise)...
+        Map<Boolean, List<Long>> resultsByValidity = clientCalls.stream()
+            .map(future -> handleFuture(future))
+            .collect(Collectors.partitioningBy(resourceId -> isValidResponse(resourceId)));
+
+        // Then the returned resource identifiers should match what is expected...
+        assertThat(resultsByValidity.getOrDefault(true, List.of()).size()).isEqualTo(successCount);
+        // And the logger mock should be called once for each exception with the expected error message
+        assertThat(resultsByValidity.getOrDefault(false, List.of()).size()).isEqualTo(errorCount);

```

Comment:
For the sake of clarity, can we have intermediate variables like validResults and invalidResults instead of this all being inline